File: php/identicon.php
<?php
    /*
    identicon.js v1.0
    maogm12@gmail.com
    */
class Identicon {
    private $text;
    private $hash;
    private $color;

    public function __construct($_text)
    {
        $this->text = $_text;
        
        // murmurhash3
        $hashVal = $this->murmurhash3($this->text);
        if ($hashVal < 0) {
            $hashVal += 0xffffffff;
        }
        $this->hash = str_pad(decbin($hashVal), 32, '0', STR_PAD_LEFT);

        //Get color
        $h = bindec(substr($this->hash, 0, 10))/1023;
        $s = bindec(substr($this->hash, 11, 10))/1023;
        $l = bindec(substr($this->hash, 22, 10))/1023;
        $this->color = $this->gen_color($h, $s, $l);
    }

    
    public function image($width = 8, $height = 8) {
        //create a white empty image using gd lib
        $im = imagecreatetruecolor($width, $height);
        $white = imagecolorallocate($im, 255, 255, 255);
        imagefilledrectangle($im, 0, 0, $width, $height, $white);
        
        //get the hash-specified color
        $im_color = imagecolorallocate($im, $this->color['r'], 
                                            $this->color['g'], 
                                            $this->color['b']);
        
        //get image-related lengthes
        $px_len = floor(min($width, $height)/8);
        $icon_edge = $px_len*8;
        $top = floor(($height - $icon_edge)/2);
        $left = floor(($width - $icon_edge)/2);

        // 01234567 for foreground
        for ($i = 0; $i < strlen($this->hash); $i++) { //should be 32 bits
            if ($this->hash[$i] === "1") {
                $xl = $left + $i%4*$px_len;
                $xr = $left + (7-$i%4)*$px_len;
                $y = $top + floor($i/4)*$px_len;
                imagefilledrectangle($im, $xl, $y, $xl+$px_len-1, $y+$px_len-1, $im_color);
                imagefilledrectangle($im, $xr, $y, $xr+$px_len-1, $y+$px_len-1, $im_color);
            }
        }
        
        ob_start();
        ImagePNG($im);      //display the image
        ImageDestroy($im);  //destroy the image
        return ob_get_clean();
    }
    
    //Helper function for HSLtoRGB
    private function HuetoRGB($v1, $v2, $vH) {
        if ( $vH < 0 ) {
            $vH += 1;
        }
        if ( $vH > 1 ) {
            $vH -= 1;
        }
        if ( ( 6 * $vH ) < 1 ) {
            return ( $v1 + ( $v2 - $v1 ) * 6 * $vH );
        }
        if ( ( 2 * $vH ) < 1 ) {
            return ( $v2 );
        }
        if ( ( 3 * $vH ) < 2 ) {
            return ( $v1 + ( $v2 - $v1 ) * ( ( 2 / 3 ) - $vH ) * 6 );
        }
        return ( $v1 );
    }
    
    //convert hsl color to rgb
    private function HSLtoRGB($H, $S, $L) {
        if ( $S == 0 ) {
            $R = $L * 255;
            $G = $L * 255;
            $B = $L * 255;
        } else {
            if ( $L < 0.5 ) {
                $var_2 = $L * ( 1 + $S );
            } else {
                $var_2 = ( $L + $S ) - ( $S * $L );
            }

            $var_1 = 2 * $L - $var_2;

            $R = 255 * $this->HuetoRGB( $var_1, $var_2, $H + ( 1 / 3 ) );
            $G = 255 * $this->HuetoRGB( $var_1, $var_2, $H );
            $B = 255 * $this->HuetoRGB( $var_1, $var_2, $H - ( 1 / 3 ) );
        }

        return array('r' => $R, 
                     'g' => $G,
                     'b' => $B);
    }

    //gen rgb color by h,s,l (both 0-1 float)
    private function gen_color($h = null, $s = null, $l = null) {
        //Generate a random nice color.'''
        if (is_null($h)) {
            //Void solid red, green, blue
            $h = rand(20, 310)/1000 + rand(0,2)/3;
        } else {
            $h = ($h-floor($h*3)/3)*0.29*3 + 0.02 + floor($h*3)/3;
        }

        if (is_null($s)) {
            //Void too dark or too bright
            $s = rand(300, 800)/100;
        } else {
            $s = $s*0.5+0.3;
        }
        
        if (is_null($l)) {
            //Void too dark or too light
            $l = rand(300, 800)/100;
        } else {
            $l = $l*0.5+0.3;
        }

        return $this->HSLtoRGB($h, $s, $l);
    }
    
    // murmurhash for php from https://github.com/lastguest/murmurhash-php
    private function murmurhash3($key, $seed = 0) {
    	$klen = strlen($key);
    	$h1   = $seed;
    	for ($i = 0, $bytes = $klen - ($remainder = $klen&3); $i < $bytes;) {
    		$k1 = ((ord($key[$i]) & 0xff))
    			| ((ord($key[++$i]) & 0xff) << 8)
    			| ((ord($key[++$i]) & 0xff) << 16)
    			| ((ord($key[++$i]) & 0xff) << 24);
    		++$i;
    		$k1  = (((($k1 & 0xffff) * 0xcc9e2d51) + (((($k1 >> 16) * 0xcc9e2d51) & 0xffff) << 16))) & 0xffffffff;
    		$k1  = $k1 << 15 | $k1 >> 17;
    		$k1  = (((($k1 & 0xffff) * 0x1b873593) + (((($k1 >> 16) * 0x1b873593) & 0xffff) << 16))) & 0xffffffff;
    		$h1 ^= $k1;
    		$h1  = $h1 << 13 | $h1 >> 19;
    		$h1b = (((($h1 & 0xffff) * 5) + (((($h1 >> 16) * 5) & 0xffff) << 16))) & 0xffffffff;
    		$h1  = ((($h1b & 0xffff) + 0x6b64) + (((($h1b >> 16) + 0xe654) & 0xffff) << 16));
    	}
    	$k1 = 0;
    	switch ($remainder) {
    		case 3: $k1 ^= (ord($key[$i + 2]) & 0xff) << 16;
    		case 2: $k1 ^= (ord($key[$i + 1]) & 0xff) << 8;
    		case 1: $k1 ^= (ord($key[$i]) & 0xff);
    		$k1  = ((($k1 & 0xffff) * 0xcc9e2d51) + (((($k1 >> 16) * 0xcc9e2d51) & 0xffff) << 16)) & 0xffffffff;
    		$k1  = $k1 << 15 | $k1 >> 17;
    		$k1  = ((($k1 & 0xffff) * 0x1b873593) + (((($k1 >> 16) * 0x1b873593) & 0xffff) << 16)) & 0xffffffff;
    		$h1 ^= $k1;
    	}
    	$h1 ^= $klen;
    	$h1 ^= $h1 >> 16;
    	$h1  = ((($h1 & 0xffff) * 0x85ebca6b) + (((($h1 >> 16) * 0x85ebca6b) & 0xffff) << 16)) & 0xffffffff;
    	$h1 ^= $h1 >> 13;
    	$h1  = (((($h1 & 0xffff) * 0xc2b2ae35) + (((($h1 >> 16) * 0xc2b2ae35) & 0xffff) << 16))) & 0xffffffff;
    	$h1 ^= $h1 >> 16;
    	return base_convert($h1,10,32);
    }
}
//Usage: $icon=new Identicon('identicon'); $icon->image(128, 128);
?>

File: python/identicon.py
'''
identicon.py
maogm12@gmail.com
Usage: python identicon.py [text]
'''
from PIL import Image, ImageDraw
import hashlib
import colorsys
import random

def murmurhash( key, seed = 0x0 ):
    ''' implements 32bit murmur3 hash. '''
    key = bytearray( key )

    def fmix( h ):
        h ^= h >> 16
        h  = ( h * 0x85ebca6b ) & 0xFFFFFFFF
        h ^= h >> 13
        h  = ( h * 0xc2b2ae35 ) & 0xFFFFFFFF
        h ^= h >> 16
        return h;

    length = len( key )
    nblocks = length / 4

    h1 = seed;

    c1 = 0xcc9e2d51
    c2 = 0x1b873593

    # body
    for block_start in xrange( 0, nblocks * 4, 4 ):
        # ??? big endian?
        k1 = key[ block_start + 3 ] << 24 | \
             key[ block_start + 2 ] << 16 | \
             key[ block_start + 1 ] <<  8 | \
             key[ block_start + 0 ]
             
        k1 = c1 * k1 & 0xFFFFFFFF
        k1 = ( k1 << 15 | k1 >> 17 ) & 0xFFFFFFFF # inlined ROTL32
        k1 = ( c2 * k1 ) & 0xFFFFFFFF;
        
        h1 ^= k1
        h1  = ( h1 << 13 | h1 >> 19 ) & 0xFFFFFFFF # inlined _ROTL32 
        h1  = ( h1 * 5 + 0xe6546b64 ) & 0xFFFFFFFF

    # tail
    tail_index = nblocks * 4
    k1 = 0
    tail_size = length & 3

    if tail_size >= 3:
        k1 ^= key[ tail_index + 2 ] << 16
    if tail_size >= 2:
        k1 ^= key[ tail_index + 1 ] << 8
    if tail_size >= 1:
        k1 ^= key[ tail_index + 0 ]
    if tail_size != 0:
        k1  = ( k1 * c1 ) & 0xFFFFFFFF
        k1  = ( k1 << 15 | k1 >> 17 ) & 0xFFFFFFFF # _ROTL32
        k1  = ( k1 * c2 ) & 0xFFFFFFFF
        h1 ^= k1

    return fmix( h1 ^ length )

def gen_color(h = None, l = None, s = None):
    '''Generate a random nice color.'''
    if h is None:
        #Void solid red, green, blue
        h = random.uniform(0.02, 0.31) + random.choice([0, 1/3.0,2/3.0])
    else:
        h = h%(1/3.0)*0.29*3+0.02 + (int(h*3)/3.0)
    
    if l is None:
        #Void too dark or too light
        l = random.uniform(0.3, 0.8)
    else:
        l = l*0.5+0.3
    
    if s is None:
        #Void too dark or too bright
        s = random.uniform(0.3, 0.8)
    else:
        s = s*0.5+0.3
    
    rgb = colorsys.hls_to_rgb(h, l, s)
    return (int(rgb[0]*256), int(rgb[1]*256), int(rgb[2]*256))

def gen_identicon(text):
    hashtext = bin(murmurhash(text, 22))[2:] #hash the text
    hashtext = hashtext.rjust(32, '0')
    size = (8,8)  #image size
    im = Image.new('RGB', size, 'white')
    draw = ImageDraw.Draw(im)

    #get the color of the icon
    h = int(hashtext[0:10], 2)/1023.0
    s = int(hashtext[11:21], 2)/1023.0
    l = int(hashtext[22:32], 2)/1023.0
    color = gen_color(h, l, s)
 
    for i in xrange(len(hashtext)): #32
        x, y = i%4, i/4
        posl = (x, y)
        posr = (7-x, y)
        if hashtext[i] == '1':
            draw.rectangle((posl, posl), fill=color)
            draw.rectangle((posr, posr), fill=color)

    del draw
    return im
    
if __name__=='__main__':
    import sys, os
    text = 'maogm12@gmail.com'
    if len(sys.argv) >= 2:
        text = sys.argv[1]
    
    im = gen_identicon(text)
    im.show()
    im.save('icon_8_8.bmp')
    im.resize((128,128)).save('icon_128_128.bmp')